plugins/samples/ab_testing/plugin.cc

```diff
@@ -0,0 +1,51 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// [START serviceextensions_plugin_ab_testing]
+#include "absl/strings/match.h"
+#include "absl/strings/str_cat.h"
+#include "proxy_wasm_intrinsics.h"
+
+constexpr std::string_view a_path = "/file1.png";
+constexpr std::string_view b_path = "/file2.png";
+constexpr int percentile = 50;
+
+class MyHttpContext : public Context {
+ public:
+  explicit MyHttpContext(uint32_t id, RootContext* root) : Context(id, root) {}
+
+  FilterHeadersStatus onRequestHeaders(uint32_t headers,
+                                       bool end_of_stream) override {
+    const auto path = getRequestHeader(":path")->view();
+
+    // Checks if the current request is eligible to be served by file B.
+    //
+    // The decision is made by hashing the request path into an integer
+    // value between 0 and 99, then comparing the hash to a predefined
+    // percentile. If the hash value is less than or equal to the percentile,
+    // the request is served by file B. Otherwise, it is served by the original
+    // file.
+    if (absl::StartsWithIgnoreCase(path, a_path) &&
+        (std::hash<std::string_view>{}(path) % 100 <= percentile)) {
```

Comment:
Hashing the entire path can result in the user experience changing between reloads for a single user because the entire query string is included in the `:path` pseudo-header. (c/o @mpwarres, Envoy makes all request look like HTTP/2 requests, and the relevant section of the H2 spec is https://datatracker.ietf.org/doc/html/rfc9113#section-8.3.1).

For example, if this was used to A/B test a page with a table, a pagination token could mess this up:

Intent: redirect data/foo to data/foov2 for 50% of users:

Reality:

1. User navigates to `example.com/data/foo?page=1&user=1234`. Their user ID hashes to the experiment group, so their URL is rewritten to `example.com/data/foov2?page=1&user=1234`.
2. User goes to the second page, `example.com/data/foo?page=2&user=1234, but this time because of `?page=2`, the path is not hashed to the experiment group, so the URL is not rewritten.
3. The user receives a poor experience where the page experience is not "sticky".

(I am being pedantic because these are canonical examples which may be copied by many users, so the importance of defaults is important).

In this case, can you only hash the user ID? (also add a test where there are other query parameters that should not change the hash behavior)

Reply:
I see... Makes sense for me.
Gonna change it following your recommendation.
Thanks for pointing this out.